lib/salestation/web/extractors.rb

```diff
@@ -192,34 +192,38 @@ def self.[](filters:, rack_key:)
         end
 
         def self.extract(filters, request_hash)
-          filters.each_with_object({}) do |filter, extracted_data|
-            case filter
-            when Symbol
-              stringified_key = filter.to_s
-              extracted_data[filter] = request_hash[stringified_key] if request_hash.key?(stringified_key)
-              if extracted_data[filter].is_a?(Array)
-                extracted_data[filter] = extracted_data[filter].map{ |hash|
-                  if hash.is_a?(Hash)
-                    hash.transform_keys(&:to_sym) if hash.is_a?(Hash)
-                  else
-                    hash
-                  end
-                }
-              end
-            when Hash
-              filter.each do |key, nested_filters|
-                stringified_key = key.to_s
-                if request_hash.key?(stringified_key)
-                  value = request_hash.fetch(stringified_key)
-                  extracted_data[key] =
-                    if !value.is_a?(Hash)
-                      value
+          # Filter as a hash is used in some existing implementations that did not expect full 
+          # recursive symbolizing of keys. In this case hash objects at the highest level of object 
+          # are represented as hash of filter keys. This is no longer needed, but we support it 
+          # to avoid a breaking change.
+
+          filters_flat = filters.map{|filter| filter.is_a?(Hash) ? filter.keys : filter}.flatten()
+
+          filters_flat.each_with_object({}) do |filter, extracted_data|
+            symbolized_key = filter 
+            stringified_key = symbolized_key.to_s 
+
+            if request_hash.key?(stringified_key)
+              extracted_data[symbolized_key] = request_hash[stringified_key]
+    
+              extracted_data[symbolized_key] =
+                case extracted_data[symbolized_key]
+                when Hash
+                  symbolized_keys = extracted_data[symbolized_key].transform_keys(&:to_sym).keys
+                  extract(symbolized_keys, extracted_data[symbolized_key])
+                when Array
+                  extracted_data[symbolized_key].map{ |elem|
+                    if elem.is_a?(Hash)
+                      symbolized_keys = elem.transform_keys(&:to_sym).keys
+                      extract(symbolized_keys, elem)
                     else
-                      extract(nested_filters, value)
+                      elem
```

Comment:
What if the array is multidimensional?
ie. elements of array  are also arrays, not hash'es

Reply:
Very good point! I have overlooked this option. I will look into it.